src/extra/gpkg/read.jl

```diff
@@ -0,0 +1,204 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+function gpkgread(fname; layer=1)
+  db = SQLite.DB(fname)
+  assertgpkg(db)
+  geoms = gpkggeoms(db; layer)
+  table = gpkgvalues(db; layer)
+  DBInterface.close!(db)
+  if eltype(table) <: Nothing
+    return georef(nothing, geoms)
+  else
+    georef(table, geoms)
+  end
+end
+
+function assertgpkg(db)
+
+  # Requirement 6: PRAGMA integrity_check returns a single row with the value 'ok'
+  # Requirement 7: PRAGMA foreign_key_check (w/ no parameter value) returns an empty result set
+  if first(DBInterface.execute(db, "PRAGMA integrity_check;")).integrity_check != "ok" ||
+     !(isempty(DBInterface.execute(db, "PRAGMA foreign_key_check;")))
+    throw(ErrorException("database integrity at risk or foreign key violation(s)"))
+  end
+
+  # Requirement 10: must include a gpkg_spatial_ref_sys table
+  # Requirement 13: must include a gpkg_contents table
+  tbcount = first(DBInterface.execute(
+    db,
+    """
+  SELECT COUNT(*) AS n FROM sqlite_master WHERE 
+  name IN ('gpkg_spatial_ref_sys', 'gpkg_contents') AND 
+  type IN ('table', 'view');
+    """
+  ))
+  if tbcount.n != 2
+    throw(ErrorException("missing required metadata tables in the GeoPackage SQL database"))
+  end
+end
+
+function gpkgvalues(db, ; layer=1)
+  table = DBInterface.execute(
+    db,
+    """
+SELECT c.table_name, c.identifier, 
+g.column_name, g.geometry_type_name, g.z, g.m, c.min_x, c.min_y, 
+c.max_x, c.max_y, 
+(SELECT type FROM sqlite_master WHERE lower(name) = 
+lower(c.table_name) AND type IN ('table', 'view')) AS object_type 
+  FROM gpkg_geometry_columns g 
+  JOIN gpkg_contents c ON (g.table_name = c.table_name)
+  WHERE 
+  c.data_type = 'features' LIMIT $layer
+    """
+  )
+  fields = nothing
+  table = map(table) do query
+    tn = query.table_name
+    tableinfo = SQLite.tableinfo(db, tn).name
+    # returns NamedTuple of AbstractVectors, also known as a "column table"
+
+    # remove `column_name` field from tableinfo in-place
+    # to avoid querying the geometry column that stores feature geometry
+    deleteat!(tableinfo, findall(x -> isequal(x, query.column_name), tableinfo))
+    columns = join(tableinfo, ", ")
+
+    # keep the shortest set of fields if there is more than one feature table
```

Comment:
I don't understand how that is possible. We are querying just the table at the given `layer`, right? The fields aren't known? Why do we need to keep track of fields across multiple tables?